zkvm/docs/zkvm-blockchain.md

```diff
@@ -0,0 +1,406 @@
+  the current time as a number of milliseconds since the Unix epoch:
+  00:00:00 UTC Jan 1,
+  1970.
+- `refscount`,
+  the number of recent block ids to cache for
+  [nonce](zkvm-spec.md#nonce)
+  uniqueness.
+
+Output:
+- a
+  [block](#block).
+
+Procedure:
+1. Compute `txroot`,
+   the root hash of an empty
+   [Merkle binary tree](zkvm-spec.md#merkle-binary-tree)
+   with label `transactions`.
+2. Compute `utxoroot`,
+   the root hash of an empty
+   [Merkle patricia tree](#merkle-patricia-tree)
+   with label `utxos`.
+3. Compute `nonceroot`,
+   the root hash of an empty
+   [Merkle patricia tree](#merkle-patricia-tree)
+   with label `nonces`.
+4. Compute `header`,
+   a
+   [block header](#block-header)
+   with its fields set as follows:
+   - `version`:
+     1
+   - `height`:
+     1
+   - `previd`:
+     all-zero string of 32-bytes
+   - `timestamp_ms`:
+     `timestamp_ms`
+   - `txroot`:
+     `txroot`
+   - `utxoroot`:
+     `utxoroot`
+   - `nonceroot`:
+     `nonceroot`
+   - `refscount`:
+     `refscount`
+5. Return a block with `header` set to `header` and an empty `txs` list.
+
+## Join existing network
+
+A new node starts here when joining a running network.
+It must either:
+- obtain all historical blocks, [applying](#apply-block) them one by one to reproduce the latest [blockchain state](#blockchain-state); or
+- obtain a recent copy of the blockchain state `state` and begin applying blocks beginning at `state.tipheader.height+1`.
+
+An obtained (as opposed to computed) blockchain state `state` may be partially validated by computing the [Merkle patricia root hash](#merkle-patricia-tree) of `state.utxos` using label `"utxos"` and verifying that it equals `state.header.utxoroot`.
+
+
+## Validate block
+
+Validating a block checks it for correctness outside the context of a particular [blockchain state](#blockchain-state).
+
+Additional correctness checks against a particular blockchain state happen during the [apply block](#apply-block) procedure.
+
+Inputs:
+- `block`,
+  the block to validate.
+- `prevheader`,
+  the header of the previous block.
+
+Output:
+- list of
+  [transaction logs](zkvm-spec.md#transaction-log),
+  one for each transaction in block.txs.
+
+Procedure:
+1. Verify `block.header.version >= prevheader.version`.
+2. Verify `block.header.height == prevheader.height+1`.
+3. Verify `block.header.previd` equals the
+   [block ID](#block-id)
+   of `prevheader`.
+4. Verify `block.header.timestamp_ms > prevheader.timestamp_ms`.
+5. Verify `block.header.refscount >= 0` and `block.header.refscount <= prevheader.refscount + 1`.
+6. Compute `txroot`, the [Merkle root hash](zkvm-spec.md#merkle-binary-tree) of `block.txs`, using label `"transactions"`.
+7. Verify `txroot == block.header.txroot`.
+8. For each transaction `tx` in block.txs:
+   1. Verify `tx.mintimems == 0` or `tx.mintimems >= block.header.timestamp_ms`.
+   2. Verify `tx.maxtimems == 0` or `tx.maxtimems < block.header.timestamp_ms`.
```

Comment:
```suggestion
   2. Verify `tx.maxtime_ms >= block.header.timestamp_ms`.
```

We allowed default-zero value in TxVM because we were using varints and were saving unspecified amount of "few bytes", so it was sort-of more convenient to allow 0 as "no limit".

In ZkVM there are no varints anywhere, so we can simplify things by using 0xffffffffffffffff as a default maxtime. It's taking up the same space as zero, and requires no special cases.